File: frontend/src/lib/ingestion.ts
// Shapes mirror the backend ingestion observability API
// (app/schemas/ingestion.py). Used by the /data page.

import { authedGet } from "@/lib/api";

export type IngestionRunStatus = "running" | "success" | "error";

export type IngestionRun = {
  id: string;
  connected_account_id: string;
  source: string;
  resource_type: string | null;
  status: IngestionRunStatus;
  started_at: string;
  finished_at: string | null;
  records_pulled: number;
  datapoints_written: number;
  error: string | null;
};

export type CountByLabel = {
  label: string;
  count: number;
};

export type IngestionStats = {
  total_datapoints: number;
  total_raw_records: number;
  by_kind: CountByLabel[];
  by_source: CountByLabel[];
  last_run_at: string | null;
};

export function listIngestionRuns(
  token: string,
  tenantId: string,
  limit = 20,
): Promise<IngestionRun[]> {
  return authedGet<IngestionRun[]>(
    `/api/v1/tenants/${tenantId}/ingestion/runs?limit=${limit}`,
    token,
  );
}

export function getIngestionStats(
  token: string,
  tenantId: string,
): Promise<IngestionStats> {
  return authedGet<IngestionStats>(
    `/api/v1/tenants/${tenantId}/ingestion/stats`,
    token,
  );
}

File: backend/app/schemas/ingestion.py
"""Read schemas for ingestion observability (the SPA "Data" page).

Read-only views over ingestion_run + datapoint for a tenant. status/kind are
exposed as plain strings to match how the ORM stores them (see models).
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class IngestionRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    connected_account_id: uuid.UUID
    source: str
    resource_type: str | None
    status: str
    started_at: datetime
    finished_at: datetime | None
    records_pulled: int
    datapoints_written: int
    error: str | None


class CountByLabel(BaseModel):
    """A single (label, count) pair for a breakdown (by kind, source, etc.)."""

    label: str
    count: int


class IngestionStats(BaseModel):
    total_datapoints: int
    total_raw_records: int
    by_kind: list[CountByLabel]
    by_source: list[CountByLabel]
    last_run_at: datetime | None
